### PR TITLE
New algorithm for DAG to RVSDG conversion

### PR DESCRIPTION
**Pull request by oflatt**
My algorithm in #413 is incorrect, as shown by the new test case.
I re-designed the algorithm and made it much simpler as a result. The new algorithm is still `O(n^2log(n))` in the number of nodes in a loop, as the previous one was.

### CODE REVIEW

File: dag_in_context/src/ast.rs [hunk 1]
```diff
@@ -42,7 +42,7 @@ pub fn val_int(i: i64) -> Value {
     Value::Const(Constant::Int(i))
 }
 
-pub fn val_state() -> Value {
+pub fn statev() -> Value {
```
Comment (oflatt) [inline dag_in_context/src/ast.rs:45]: rename for consistancy

File: src/rvsdg/dag_region_query.rs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                for child in children {
+                    stack.push(child);
+                }
+            }
+        }
+
+        result
+    }
+
+    pub(crate) fn get(&self, expr: &RcExpr) -> HashSet<*const Expr> {
```
Comment (ajpal) [inline src/rvsdg/dag_region_query.rs:63]: What does this get?

File: src/rvsdg/dag_region_query.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    // regardless of branching
+    always_executed: HashMap<*const Expr, HashSet<*const Expr>>,
+}
+
+impl AlwaysExecutedCache {
+    /// Given a conditional expression, finds all children that are executed regardless of
+    /// the branch taken.
+    /// Once a node is added to the result, children of that node are not added (since any node it
+    /// depends on will be executed).
+    pub(crate) fn always_executed(
```
Comment (ajpal) [inline src/rvsdg/dag_region_query.rs:23]: Rename - the name here should have something to do with not-the-children

File: src/rvsdg/dag_region_query.rs (hunk 2 above)
Comment (ajpal) [reply src/rvsdg/dag_region_query.rs:63]: (add doc comment)

File: src/rvsdg/dag_region_query.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+                    }
+
+                    res.extend(branches_intersection);
+                    res
+                }
+                _ => {
+                    let children = expr.children_same_scope();
+                    let mut res = HashSet::new();
+                    for (i, child) in children.iter().enumerate() {
+                        if i == 0 {
```
Comment (ajpal) [inline src/rvsdg/dag_region_query.rs:97]: why special case for i = 0?

File: src/rvsdg/dag_region_query.rs [hunk 5]
```diff
@@ -0,0 +1,195 @@
+//! This file is a helper for translation from the dag IR to RVSDGs.
+//! It contains the `AlwaysExecutedCache` struct, which is used to
+//! query nodes that are guaranteed to be executed given that a particular node is executed.
+//! This information is used by `from_dag.rs` to compute the input nodes to branch regions.
+
+use std::rc::Rc;
+
+use dag_in_context::schema::{Expr, RcExpr};
+use hashbrown::{HashMap, HashSet};
```
[the remark this answers is not in the document]
Comment (oflatt) [reply src/rvsdg/dag_region_query.rs:9]: fixed to use std

**Review by ajpal — APPROVED** [review]
great work!

**Pull request merged**